### DISCUSSION TITLE
PGroongaを活用した前方一致検索と完全一致検索について

### DISCUSSION THREAD

**Discussion by nakayman** [pgroonga/pgroonga#361] [Q&A]
初めて投稿させていただきます。
英字項目に対してPGroongaを活用した前方一致検索と完全一致検索で同じ結果が得られる事象が発生しており、
原因究明のための手がかりとなる情報を頂きたい次第でございます。

■前知識
・実行SQL:検索ワード)assem
前方一致検索：
`select eigo from test.tabel1 where ((eigof || '') &@~ '\|assem' OR (eigof || '' ||'') &@~ '\|assem')`

完全一致検索：
`select eigo from test.tabel1 where ((eigof || '') &@~ '\|assem\|' OR (eigof || '' ||'') &@~ '\|assem\|')`

※eigofは「|」で囲まれた状態で格納されている

完全一致検索の想定結果としましては、|assem|の該当0件でしたが、
前方一致検索と完全一致検索で同じ結果が得られています。
検索結果例）
assembly
assembler
assembly language
assembling 
assemblage

適用されているインデックス：
`CREATE INDEX idx_pgroonga_stem ON test.tabel1 USING pgroonga ((eigof || '')) WITH (plugins='token_filters/stem', tokenizer='TokenBigramSplitSymbol', token_filters='TokenFilterStem')`

`CREATE INDEX idx_pgroonga_bigram ON test.tabel1 USING pgroonga ((eigof || '' || '')) WITH (tokenizer='TokenBigramSplitSymbolAlphaDigit')`

■質問事項
完全一致で検索する際に「\|assem\|」のように「|」で閉じた状態で実行しますが、語尾の「|」が無視されているような挙動になっていますが、
これはトークナイザー「TokenBigramSplitSymbol」によってステミング機能が有効になっているからでしょうか。

**Comment by kou** [comment]
「マッチエスカレーション」と呼んでいる「ヒットしなかったときに自動でゆるくヒットする機能」が発動しているからです。

↓でこの機能を無効にしてみてください。

```SQL
SET pgroonga.match_escalation_threshold = -1;
```

参考： https://pgroonga.github.io/ja/reference/parameters/match-escalation-threshold.html

**Reply by nakayman** [reply]
@kou 様
早速のご対応ありがとうございます。試してみます。
